keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -907,22 +917,26 @@ Problem examples:
 | CRD is in multiple clusters, but schema differs        | If policy author is aware of the schema variations, can they write policies that work for all the variations? |
 | Validation of an aggregated API server type            | Main API server does not have type definitions |
 
-Due to these complications, we have decided to evalute CEL expressions
-dynamically. Informational type checking will be provided (except for aggregated
-API server types), but will be surfaced only as warnings. See "Alternatives
-Considered" section for details of all the alternatives we reviewed when
-selecting this approach.
+Due to these complications, Type checking will be optional which can be activated by an expression-level
+`TypeCheckingMode` field. Possible modes are:
 
-Type checking is still performed for all expressions where a GVK can be matched
-to type check against, resulting in warnings, e.g.:
+- `Never` Do not type check the referred object or the param any time. 
+  Treat `object`, `oldObject`, and `params` as dynamic-typed unstructured objects. 
+  This behavior matches that of Kubernetes 1.26
+- `Evaluation` Type check the referred `object`, `oldObject`, and `params` when the expression is being compiled and evaluated after the policy is created. 
+  If the type check fails, mark the policy as misconfigured. Do not type check when the policy is being created or updated.
+- `Always` Type check the referred `object`, `oldObject`, and `params` immediately when the policy is being created or updated. 
+  If the type check fails, including the situation where the schema fails to resolve, the API server will reject the operation on the policy object. 
+  The same type check will happen during compilation and evaluation.
+
+
+The type check may fail in one of the three reasons:
+1. The schema is resolved and parsed. The CEL library determines the expression is invalid;
+2. The schema is resolved, but cannot be parsed. This is considered an internal error and will be logged;
+3. The schema cannot be resolved because the referred type does not yet exist. This should be a transient error.
+
+If TypeCheckingMode is set to Evaluation, and either 2 or 3 happens, the compilation will fall back to unstructured as if TypeCheckingMode was set to Never. This allows the evaluation to continue before the referred CRDs are created or published.
```

Comment:
Parsed expressions run equivalently to checked expressions for the standard library, but they're slower. Since all of your custom functions are bound using e.g. `cel.Function()`, then the bindings will handle dynamic dispatch (what happens on parse vs. checked exprs) to various overloads automatically as well.

Reply:
Thank you both for the information. I did know `Compile` means `Parse` then `Check`. Currently, we only use `Compile`. I can do some testing on how splitting these two stages would help.

@TristonianJones quick question. What would be difference of evaluating unchecked expression vs checked  but everything is `dyn` type. If everything is `dyn` then the dynamic dispatch problem would still exist. As of the past release (1.26), we treated everything as `dyn`.

Reply:
@jiahuif There's no change in evaluation semantics, just internal overhead to resolve which function to invoke.

Reply:
Good to know. Thank you!